orangecontrib/text/widgets/owcorpusviewer.py

```diff
@@ -118,28 +125,65 @@ def copy_to_clipboard(self):
         QApplication.clipboard().setText(text)
 
     def set_data(self, data=None):
+        self.closeContext()
         self.reset_widget()
         self.corpus = data
+        self.search_features = []
         if data is not None:
             if not isinstance(data, Corpus):
                 self.corpus = Corpus.from_table(data.domain, data)
-            self.load_features()
+            domain = self.corpus.domain
+
+            self.search_features = list(filter_visible(chain(domain.variables, domain.metas)))
+            # set names here, openContext will map them to features
+            self.display_features = [f.name for f in filter_visible(chain(domain.variables, domain.metas))]
+            self.search_indices = list(range(len(self.search_features)))
+            self.display_indices = list(range(len(self.display_features)))
+            self.openContext(self.corpus)
             self.regenerate_docs()
+            self.list_docs()
+            self.update_info()
+            self.set_selection()
+            self.show_docs()
+            # Enable/disable tokens checkbox
+            if not self.corpus.has_tokens():
+                self.show_tokens_checkbox.setCheckState(False)
+            self.show_tokens_checkbox.setEnabled(self.corpus.has_tokens())
         self.commit()
 
+    def closeContext(self):
+        """ Workaround to store a list of feature names instead of features
+        itself since storing a list of features doesn't yet work. """
+
+        # assigning to display_features deletes indices, hence make a copy
+        indices = self.display_indices[:]
+        # map features to name
+        self.display_features = [f.name for f in self.display_features]
+        # restore selected indices
+        self.display_indices = indices
+        super().closeContext()
+
+    def openContext(self, data):
+        """ Revert features names back to features and keep the indices. """
+        super().openContext(data)
+        domain = self.corpus.domain
```

Comment:
openContext can potentially work also with domain given as data. If we restrict ourselves to always passing the data we can refactor to your suggestion.